src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,807 @@
+                    // Contains a canceled bool to indicate if the send was canceled.
+                    case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                        {
+                            status = HandleEventSendComplete(ref evt);
+                        }
+                        break;
+                    // Peer has told us to shutdown the reading side of the stream.
+                    case QUIC_STREAM_EVENT.PEER_SEND_SHUTDOWN:
+                        {
+                            status = HandleEventPeerSendShutdown();
+                        }
+                        break;
+                    // Peer has told us to abort the reading side of the stream.
+                    case QUIC_STREAM_EVENT.PEER_SEND_ABORTED:
+                        {
+                            status = HandleEventPeerSendAborted();
+                        }
+                        break;
+                    // Peer has stopped receiving data, don't send anymore.
+                    // Potentially throw when WriteAsync/FlushAsync.
+                    case QUIC_STREAM_EVENT.PEER_RECEIVE_ABORTED:
+                        {
+                            status = HandleEventPeerRecvAbort();
+                        }
+                        break;
+                    // Occurs when shutdown is completed for the send side.
+                    // This only happens for shutdown on sending, not receiving
+                    // Receive shutdown can only be abortive.
+                    case QUIC_STREAM_EVENT.SEND_SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventSendShutdownComplete(ref evt);
+                        }
+                        break;
+                    // Shutdown for both sending and receiving is completed.
+                    case QUIC_STREAM_EVENT.SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventShutdownComplete();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicStatusCodes.InternalError;
+            }
+
+            return status;
+        }
+
+        private unsafe uint HandleEventRecv(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            StreamEventDataRecv receieveEvent = evt.Data.Recv;
+            for (int i = 0; i < receieveEvent.BufferCount; i++)
+            {
+                _receiveQuicBuffers.Add(receieveEvent.Buffers[i]);
+            }
+
+            bool shouldComplete = false;
+            lock (_sync)
+            {
+                if (_readState == ReadState.None)
+                {
+                    shouldComplete = true;
+                }
+                _readState = ReadState.IndividualReadComplete;
+            }
```

Comment:
More efficient:

```c#
bool shouldComplete = Interlocked.Exchange(_readState, ReadState.IndividualReadComplete) == ReadState.None;
```

Reply:
Maybe, but it would make other checks much more complicated. For example, in HandleEventShutdownComplete, I'm updating both the read state and shutdown state under the sync. 

```c#
 lock (_sync)
            {
                // This event won't occur within the middle of a receive.
                if (NetEventSource.IsEnabled) NetEventSource.Info("Completing resettable event source.");

                if (_readState == ReadState.None)
                {
                    shouldReadComplete = true;
                }

                _readState = ReadState.ReadsCompleted;

                if (_shutdownState == ShutdownWriteState.None)
                {
                    _shutdownState = ShutdownWriteState.Finished;
                    shouldShutdownWriteComplete = true;
                }
            }
```

Doing interlocked exchanges could introduce interleavings that we haven't fully thought through. 

I can leave it as a TODO though; it may be worth the investment if the sync lock is too hot.